### ISSUE TITLE
Tracking issue for future-incompatibility lint `ptr_cast_add_auto_to_object`

### ISSUE THREAD

**Issue by WaffleLapkin** [rust-lang/rust#127323]
This is a tracking issue for the `ptr_cast_add_auto_to_object` lint, which was added in #120248.

This lint detects casts of raw pointers to trait objects, which add auto traits. Adding auto traits to trait objects may cause UB when [`#![feature(arbitrary_self_types)]`](https://github.com/rust-lang/rust/issues/44874) is used.

## Example

```rust
#![feature(arbitrary_self_types)]
trait Trait {
    fn f(self: *const Self)
    where
        Self: Send;
}

impl Trait for *const () {
    fn f(self: *const Self) {
        unreachable!()
    }
}

fn main() {
    let unsend: *const () = &();
    let unsend: *const dyn Trait = &unsend;
    let send_bad: *const (dyn Trait + Send) = unsend as _;
    send_bad.f(); // this crashes, since vtable for `*const ()` does not have an entry for `f`
    //~^ warning: adding an auto trait `Send` to a trait object in a pointer cast may cause UB later on
}
```

In case your usage is sound (e.g. because the trait doesn't have auto trait bounds), you can replace cast with a transmute to suppress the warning:

```rust
trait Cat {}
impl Cat for *const () {}

fn main() {
    let unsend: *const () = &();
    let unsend: *const dyn Cat = &unsend;
    let _send: *const (dyn Cat + Send) = unsafe {
        // Safety:
        // - Both types are pointers, to the same trait object (and thus have the same vtable)
        // - `Cat` does not have methods with `Send` bounds
        std::mem::transmute::<*const dyn Cat, *const (dyn Cat + Send)>(unsend)
    };
    // meow
}
```

**Comment by bjorn3** [comment]
> In case your usage is sound (e.g. because the trait doesn't have auto trait bounds), you can replace cast with a transmute to suppress the warning:

Future compatibility warnings are generally made hard errors after a while, right? I don't think we should suggest suppressing a future compatibility warning.

**Comment by WaffleLapkin** [comment]
@bjorn3 we only will make cases with the warning into errors. `transmute`ing is, and will be, allowed (although you _do_ actually need to make sure that your code is sound).

**Comment by RalfJung** [comment]
We already have a safety invariant of requiring a valid vtable for raw pointers; dyn upcasting relies on that. I guess the cast above is accepted because "marker traits don't change the vtable", or so?

**Comment by RalfJung** [comment]
@BoxyUwU that does *not* trigger the lint that this issue tracks... so it should probably get a separate issue?

**Comment by BoxyUwU** [comment]
The ICE gets hit before the lint is triggered but once the ICE is fixed the lint triggers + there's a segfault when the program is run

**Comment by BoxyUwU** [comment]
```
boxy@crossaints:/media/Nyoomies/Repos/playground_1$ cargo +stage1 run
warning: adding an auto trait `Send` to a trait object in a pointer cast may cause UB later on
  --> src/main.rs:29:13
   |
29 |     let a = a as *const dyn Trait as *const (dyn Trait + Send);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #127323 <https://github.com/rust-lang/rust/issues/127323>
   = note: `#[warn(ptr_cast_add_auto_to_object)]` on by default

warning: `c` (bin "c") generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.00s
warning: the following packages contain code that will be rejected by a future version of Rust: c v0.1.0 (/media/Nyoomies/Repos/playground_1)
note: to see what the problems were, use the option `--future-incompat-report`, or run `cargo report future-incompatibilities --id 2`
     Running `target/debug/c`
Segmentation fault (core dumped)
```
If anyone is curious this is the output I get locally on that example when using a stage1 compiler from errs' branch

**Comment by RalfJung** [comment]
Ah okay. So this one "just" needs the lint to be turned into a hard error then.

I agree with your statement elsewhere that that should happen before this example works on stable.

**Comment by Manishearth** [comment]
> ```rust
>         // Safety:
>         // - Both types are pointers, to the same trait object (and thus have the same vtable)
>         // - `Cat` does not have methods with `Send` bounds
>         std::mem::transmute::<*const dyn Cat, *const (dyn Cat + Send)>(unsend)
> ```


I'm actually surprised this is sound: should this be documented more? As an unsafe reviewer this code would set off major alarm bells for me.

I can kind of piece together why this works but I don't think this particular behavior of auto traits on trait objects is actually documented.

**Comment by Manishearth** [comment]
This also feels to me to be a bug with object safety: a trait with `where Self: Send` methods probably shouldn't be object safe in the first place, if trait objects and auto traits work this way? I can't fully tell if it's _computable_ that a trait has a potential for this problem, but it feels cleaner and more clear to users if the outcome here were that making `dyn Trait` ended up forcing you to add `+ Sized` to `where Self: Send`.

**Comment by RalfJung** [comment]
@Manishearth why would such a method be a problem for dyn compatibility / object safety?

**Comment by Manishearth** [comment]
@RalfJung Maybe I've misunderstood the issue, but it seems like a trait method with `where Self: Send` cannot show up on trait objects. Typically, when we have trait methods that cannot show up on trait objects, we consider that an object safety violation; e.g. we do not silently hide `fn foo(&self, x: Self)` from trait objects, and you are required to explicitly acknowledge their unavailability on objects via a `where Self: Sized` on the method.

That seems to be the same principle that gets applied here?

**Comment by BoxyUwU** [comment]
The general case of `Self: Trait` bounds on trait methods does cause the trait to not be dyn compatible. This didn't used to be the case #51443 and has since been fixed. `Self: AutoTrait` was deliberately left out of that lint in #107082 to avoid breaking https://github.com/dtolnay/async-trait.

`Self: AutoTrait` doesn't cause any problems on dyn compatible traits as you cannot `impl AutoTrait for dyn Trait` so it is always true that dyn types only implement auto traits if the underlying type does as well. Before raw pointers as receivers this was perfectly coherently logic (afaik) as even if you could cast `*const dyn Trait` to `*const dyn Trait + Send` you would only be able to dispatch to it by converting it to a reference (or smart pointer from std which requires unsafe) which would not be sound in general anyway.

> I don't think this particular behavior of auto traits on trait objects is actually documented.

I don't think it is yeah, in general the documentation of dyn compatibility rules is not done yet but it's on my todo list 🤔 #127251

**Comment by RalfJung** [comment]
I would expect that if I have `x: dyn Trait + Send`, I can call a `Self: Send` method, and otherwise I cannot. That seems entirely coherent to me with the idea that `Self` *is* `dyn Trait` when invoking a method through `dyn Trait`.

So yeah I guess it's more like the `Self: Sized` case than the normal `Self: Trait` case.

**Comment by traviscross** [comment]
We made this into a hard error in:

- https://github.com/rust-lang/rust/pull/136764

I've checked, and there don't seem to be any references to this issue in the compiler (other than registering the removed lint).  It seems, then, that we can close this.

For overall tracking on the set of changes we've made to make pointer casts of trait objects stricter, we'll use:

- https://github.com/rust-lang/rust/issues/141402
